### PR TITLE
Replace CSV Ragel parser by a hand-written parser copied from encoding/csv

### PR DESCRIPTION
**Pull request by kennytm**
### What problem does this PR solve? 

The Ragel-based CSV parser is much slower than the standard `encoding/csv` parser.

As explained in #111, the parser was using Ragel to reuse existing framework for simplicity. However as more and more customer start to use Lightning with CSV input, this is the time we need to optimize the things.

### What is changed and how it works?

The new parser was inspired by `encoding/csv` but almost nothing remained except the [`recordBuffer`/`fieldIndexes` members](https://github.com/golang/go/blob/a2bff7c2964c6bf2c9741eb767d749d773f20770/src/encoding/csv/reader.go#L155-L163) to reduce the amount of allocations. We cannot reuse `encoding/csv` because:

* `encoding/csv` does not allow us to track the current read position
* `encoding/csv` does not recognize backslash-escaped fields (required for MySQL-generated CSV)
* `encoding/csv` does not support *disabling* quoting 

Implementing these make the new parser still worse than `encoding/csv`, but much better than the original one.

| Parser | TPC-C "CUSTOMER" Row |
|--------|---------------------:|
| `encoding/csv` | 1898 ns/op |
| Original parser | 6930 ns/op |
| New parser | 2426 ns/op |

So the new parser is 35% of the original parser and `encoding/csv` is 80% of the new parser. We can investigate how to squeeze out the remaining 20% later.

### Check List 

Tests 

 - Unit test

Side effects

 - Increased code complexity

Related changes

 - Need to be included in the release note
    * Improved CSV parsing speed by (TO BE FILLED IN)%

### CODE REVIEW

**Comment by siddontang** [comment]
do we have a benchmark test for lightning, e.g. 10GB data, 3 TiKV, running `/bench`, @zhouqiang-cl

**Comment by zhouqiang-cl** [comment]
> do we have a benchmark test for lightning, e.g. 10GB data, 3 TiKV, running `/bench`, @zhouqiang-cl

Now. we did not have benchmark for lightning. will scheduler for it

File: lightning/mydump/csv_parser.go [hunk 1]
```diff
@@ -10,143 +24,335 @@ import (
...
+			if len(records[i-1]) > 0 {
+				break
 			}
-			return nil
+			i--
 		}
+		records = records[:i]
 	}
+
+	row.Row = make([]types.Datum, 0, len(records))
```
Comment (XuHuaiyu) [inline lightning/mydump/csv_parser.go]: It will be better to use sync.Pool here

**Comment by kennytm** [comment]
PTAL again @XuHuaiyu 

```
before recycling:

PASS: csv_parser_test.go:690: benchCSVParserSuite.BenchmarkReadRowUsingMydumpCSVParser   1000000              2875 ns/op            2144 B/op          3 allocs/op

after recycling:

PASS: csv_parser_test.go:690: benchCSVParserSuite.BenchmarkReadRowUsingMydumpCSVParser    500000              2415 ns/op             899 B/op          3 allocs/op
```

File: lightning/mydump/csv_parser.go [hunk 2]
```diff
@@ -10,143 +24,335 @@ import (
...
-			if !parser.cfg.TrimLastSep {
-				parser.appendEmptyValues(emptySepCount)
+	records, err := parser.readRecord()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	// remove trailing empty values
+	if parser.cfg.TrimLastSep {
+		i := len(records)
+		for i > 0 {
```
Comment (july2993) [inline lightning/mydump/csv_parser.go]: suggest change for statement as:
```
for i > 0 && len(records[i-1]) <= 0  {
    i--
}
```

File: lightning/mydump/csv_parser_test.go [hunk 3]
```diff
@@ -1144,15 +647,103 @@ func (s *testMydumpCSVParserSuite) TestSyntaxErrorLog(c *C) {
...
+			parser.RecycleRow(parser.LastRow())
+			rowsCount++
+			continue
+		}
+		if errors.Cause(err) == io.EOF {
+			return
+		}
+		c.Fatal(err)
+	}
+	c.Assert(rowsCount, Equals, c.N)
```
Comment (july2993) [inline lightning/mydump/csv_parser_test.go]: unreachable code

File: lightning/mydump/csv_parser_test.go [hunk 4]
```diff
@@ -1144,15 +647,103 @@ func (s *testMydumpCSVParserSuite) TestSyntaxErrorLog(c *C) {
...
+			datums = datums[:0]
+			rowsCount++
+			continue
+		}
+		if errors.Cause(err) == io.EOF {
+			return
+		}
+		c.Fatal(err)
+	}
+	c.Assert(rowsCount, Equals, c.N)
```
Comment (july2993) [inline lightning/mydump/csv_parser_test.go]: unreachable code

**Comment by kennytm** [comment]
/run-all-tests tidb=release-3.1 pd=release-3.1 tikv=release-3.1

**Comment by july2993** [comment]
fail to start tidb？ 🤔
```
[2020-03-13T21:46:01.463Z] [2020/03/14 05:45:58.739 +08:00] [WARN] [base_client.go:170] ["[pd] failed to get cluster id"] [url=http://127.0.0.1:2379] [error="tls: found a certificate rather than a key in the PEM for the private key"] [errorVerbose="tls: found a certificate rather than a key in the PEM for the private key\ngithub.com/pingcap/pd/pkg/grpcutil.SecurityConfig.ToTLSConfig\n\t/home/jenkins/agent/workspace/tidb_release-3.1/go/pkg/mod/github.com/pingcap/pd@v1.1.0-beta.0.20200213133706-fbbe75e180e6/pkg/grpcutil/grpcutil.go:49\ngithub.com/pingcap/pd/client.(*baseClient).getOrCreateGRPCConn\n\t/home/jenkins/agent/workspace/tidb_release-3.1/go/pkg/mod/github.com/pingcap/pd@v1.1.0-beta.0.20200213133706-fbbe75e180e6/client/base_client.go:264\ngithub.com/pingcap/pd/client.(*baseClient).getMembers\n\t/home/jenkins/agent/workspace/tidb_release-3.1/go/pkg/mod/github.com/pingcap/pd@v1.1.0-beta.0.20200213133706-fbbe75e180e6/client/base_client.go:202\ngithub.com/pingcap/pd/client.(*baseClient).initClusterID\n\t/home/jenkins/agent/workspace/tidb_release-3.1/go/pkg/mod/github.com/pingcap/pd@v1.1.0-beta.0.20200213133706-fbbe75e180e6/client/base_client.go:167\ngithub.com/pingcap/pd/client.(*baseClient).initRetry\n\t/home/jenkins/agent/workspace/tidb_release-3.1/go/pkg/mod/github.com/pingcap/pd@v1.1.0-beta.0.20200213133706-fbbe75e180e6/client/base_client.go:103\ngithub.com/pingcap/pd/client.newBaseClient\n\t/home/jenkins/agent/workspace/tidb_release-3.1/go/pkg/mod/github.com/pingcap/pd@v1.1.0-beta.0.20200213133706-fbbe75e180e6/client/base_client.go:84\ngithub.com/pingcap/pd/client.NewClientWithContext\n\t/home/jenkins/agent/workspace/tidb_release-3.1/go/pkg/mod/github.com/pingcap/pd@v1.1.0-beta.0.20200213133706-fbbe75e180e6/client/client.go:136\ngithub.com/pingcap/pd/client.NewClient\n\t/home/jenkins/agent/workspace/tidb_release-3.1/go/pkg/mod/github.com/pingcap/pd@v1.1.0-beta.0.20200213133706-fbbe75e180e6/client/client.go:130\ngithub.com/pingcap/tidb/store/tikv.Driver.Open\n\t/home/jenkins/agent/workspace/tidb_release-3.1/go/src/github.com/pingcap/tidb/store/tikv/kv.go:82\ngithub.com/pingcap/tidb/store.newStoreWithRetry.func1\n\t/home/jenkins/agent/workspace/tidb_release-3.1/go/src/github.com/pingcap/tidb/store/store.go:70\ngithub.com/pingcap/tidb/util.RunWithRetry\n\t/home/jenkins/agent/workspace/tidb_release-3.1/go/src/github.com/pingcap/tidb/util/misc.go:51\ngithub.com/pingcap/tidb/store.newStoreWithRetry\n\t/home/jenkins/agent/workspace/tidb_release-3.1/go/src/github.com/pingcap/tidb/store/store.go:68\ngithub.com/pingcap/tidb/store.New\n\t/home/jenkins/agent/workspace/tidb_release-3.1/go/src/github.com/pingcap/tidb/store/store.go:52\nmain.createStoreAndDomain\n\t/home/jenkins/agent/workspace/tidb_release-3.1/go/src/github.com/pingcap/tidb/tidb-server/main.go:207\nmain.main\n\t/home/jenkins/agent/workspace/tidb_release-3.1/go/src/github.com/pingcap/tidb/tidb-server/main.go:172\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:203\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1357\ngithub.com/pingcap/pd/client.(*baseClient).getOrCreateGRPCConn\n\t/home/jenkins/agent/workspace/tidb_release-3.1/go/pkg/mod/github.com/pingcap/pd@v1.1.0-beta.0.20200213133706-fbbe75e180e6/client/base_client.go:266\ngithub.com/pingcap/pd/client.(*baseClient).getMembers\n\t/home/jenkins/agent/workspace/tidb_release-3.1/go/pkg/mod/github.com/pingcap/pd@v1.1.0-beta.0.20200213133706-fbbe75e180e6/client/base_client.go:202\ngithub.com/pingcap/pd/client.(*baseClient).initClusterID\n\t/home/jenkins/agent/workspace/tidb_release-3.1/go/pkg/mod/github.com/pingcap/pd@v1.1.0-beta.0.20200213133706-fbbe75e180e6/client/base_client.go:167\ngithub.com/pingcap/pd/client.(*baseClient).initRetry\n\t/home/jenkins/agent/workspace/tidb_release-3.1/go/pkg/mod/github.com/pingcap/pd@v1.1.0-beta.0.20200213133706-fbbe75e180e6/client/base_client.go:103\ngithub.com/pingcap/pd/client.newBaseClient\n\t/home/jenkins/agent/workspace/tidb_release-3.1/go/pkg/mod/github.com/pingcap/pd@v1.1.0-beta.0.20200213133706-fbbe75e180e6/client/base_client.go:84\ngithub.com/pingcap/pd/client.NewClientWithContext\n\t/home/jenkins/agent/workspace/tidb_release-3.1/go/pkg/mod/github.com/pingcap/pd@v1.1.0-beta.0.20200213133706-fbbe75e180e6/client/client.go:136\ngithub.com/pingcap/pd/client.NewClient\n\t/home/jenkins/agent/workspace/tidb_release-3.1/go/pkg/mod/github.com/pingcap/pd@v1.1.0-beta.0.20200213133706-fbbe75e180e6/client/client.go:130\ngithub.com/pingcap/tidb/store/tikv.Driver.Open\n\t/home/jenkins/agent/workspace/tidb_release-3.1/go/src/github.com/pingcap/tidb/store/tikv/kv.go:82\ngithub.com/pingcap/tidb/store.newStoreWithRetry.func1\n\t/home/jenkins/agent/workspace/tidb_release-3.1/go/src/github.com/pingcap/tidb/store/store.go:70\ngithub.com/pingcap/tidb/util.RunWithRetry\n\t/home/jenkins/agent/workspace/tidb_release-3.1/go/src/github.com/pingcap/tidb/util/misc.go:51\ngithub.com/pingcap/tidb/store.newStoreWithRetry\n\t/home/jenkins/agent/workspace/tidb_release-3.1/go/src/github.com/pingcap/tidb/store/store.go:68\ngithub.com/pingcap/tidb/store.New\n\t/home/jenkins/agent/workspace/tidb_release-3.1/go/src/github.com/pingcap/tidb/store/store.go:52\nmain.createStoreAndDomain\n\t/home/jenkins/agent/workspace/tidb_release-3.1/go/src/github.com/pingcap/tidb/tidb-server/main.go:207\nmain.main\n\t/home/jenkins/agent/workspace/tidb_release-3.1/go/src/github.com/pingcap/tidb/tidb-```

**Comment by 3pointer** [comment]
this [`seek`](https://github.com/pingcap/tidb-lightning/blob/master/lightning/mydump/region.go#L224) seems unnecessary after this PR changed

**Comment by july2993** [comment]
@kennytm pls resolve conflicts

**Comment by july2993** [comment]
/run-all-test
seems `tests/check_requirements/` fail by chance

**Comment by kennytm** [comment]
PTAL @3pointer

**Pull request merged**